### PR TITLE
Implement GObject properties

### LINKED ISSUE #43: Support to declare GObject properties
This is meant to be a discussion about the implementation of user declared GObject properties.

IMO it must be as transparent as possible, i.e. a `property`/`setter`/`getter` macro on GObject, overwriting the original Crystal Object macros.

```Crystal
class Foo < GObject::Object
  # Creates a read/write GObject property
  property foo : Int32
  # Creates a read only GObject property
  getter bar : Int32
  # Creates a write only GObject property
  setter moo : Int32 
  # Prop with blurb
  property hey : String, blurb: "I'm the property blurb"
  # Prop with cusom nick
  property nicked : String, nick: "Nickey"
end
```
The nick can be the name.titleize, or custom. Blurb could be default to name.titleize too

These macros would create the variables on Crystal world, so people could use `@foo` to read `foo` property without overhead.

A possible implementation for this would be the macro generate something like

```Crystal
@[GObject::Property(blurb: "Prop blurb"]
@foo : String
```

So a finalize macro would catch all instance variables with this annotation and generate the set_property/get_property methods plus the class_init that register the properties.

I didn't test this yet, but if it works we could use the same schema to vfuncs, to avoid the bug that forbid a method starting with `do_` to be declared if no vfuncs with that name exists, so instead vfuncs can just be:

```
@[GObject::VFunc]
def do_whatever

# or

@[GObject::VFunc(name: :whatever)]
def whatever_name
```

But this is another subject :grin: 

Back to properties... other options for properties can also be added as extra parameters in the macro, e.g.:

```Crystal
property foo : Int32, min: 0, max: 32
```

### PR DESCRIPTION
**Pull request by BlobCodes**
Closes #43 

Syntax:
```crystal
private class UserObjectWithGProperties < GObject::Object
  @[GObject::Property(nick: "STRING", blurb: "A string without meaning", default: "default")]
  property str_ing : String = "default"

  @[GObject::Property(nick: "INTEGER", blurb: "An Int32", default: 42, min: 40, max: 50)]
  property int : Int32 = 42

  @[GObject::Property(default: TestFlags::BC)]
  property flags = TestFlags::BC

  @[GObject::Property]
  property object : UserObject? = nil
end
```

- A property is readable if a getter for the value exists (`def property_name`).
- A property is readable if a setter for the value exists (`def property_name=`).
- `GObject::Object` properties must be nilable because they can be set to `null` from C.
- All options inside the `GObject::Property` annotation are optional.

Beyond the properties themselves, I also changed a few bits elsewhere:
- Split some methods in `GValue` into class method / instance method
- Stopped assuming `g_value_get_object` is never `null`

### CODE REVIEW

**Comment by BlobCodes** [comment]
But anyways it seems to work in the gtk inspector:
Bildschirmfoto vom 2022-06-22 21-50-40

This will be very useful for binding properties directly in gtk templates.

File: spec/inheritance_spec.cr [hunk 1]
```diff
@@ -90,4 +107,86 @@ describe "Classes inheriting GObject::Object" do
...
   end
+
+  it "can set GObject properties" do
+    obj = UserObjectWithGProperties.new
+
+    out_string = LibGObject::Value.new
+    LibGObject.g_value_init(pointerof(out_string), GObject::TYPE_STRING)
+    LibGObject.g_value_set_string(pointerof(out_string), "test value")
+    LibGObject.g_object_set_property(obj, "str-ing", pointerof(out_string))
+    LibGObject.g_value_unset(pointerof(out_string))
```
Comment (hugopl) [inline spec/inheritance_spec.cr:118]: You can use [g_object_set](https://docs.gtk.org/gobject/method.Object.set.html) to set the property and avoid dealing with GValue at all, GICrystal currently uses it to set properties.
```suggestion
    LibGObject.g_object_set(obj, "str-ing", "test value", Pointer(Void).null)
```

File: spec/inheritance_spec.cr [hunk 2]
```diff
@@ -90,4 +107,86 @@ describe "Classes inheriting GObject::Object" do
     LibGObject.g_type_check_instance_is_a(obj, UserSubject.g_type).should eq(1)
     obj.string.should eq("hey")
   end
+
+  it "can set GObject properties" do
```
Comment (hugopl) [inline spec/inheritance_spec.cr:111]: Better to create a `properties_spec.cr` file with the properties spec instead of put the tests in inheritance specs.

Comment (hugopl) [reply spec/inheritance_spec.cr:111]: This test is testing if you can get the value of Crystal declared GObject properties, i.e. you set them _"in C"_ and get them in Crystal, so I think the test case names are swapped.

File: spec/inheritance_spec.cr [hunk 3]
```diff
@@ -90,4 +107,86 @@ describe "Classes inheriting GObject::Object" do
...
+    obj = UserObjectWithGProperties.new
+
+    out_string = LibGObject::Value.new
+    LibGObject.g_object_get_property(obj, "str-ing", pointerof(out_string))
+    GObject::Value.raw(GObject::TYPE_STRING, pointerof(out_string).as(Void*)).as(String?).should eq("default")
+    LibGObject.g_value_unset(pointerof(out_string))
+    obj.str_ing = "test value"
+    LibGObject.g_object_get_property(obj, "str-ing", pointerof(out_string))
+    GObject::Value.raw(GObject::TYPE_STRING, pointerof(out_string).as(Void*)).as(String?).should eq("test value")
+    LibGObject.g_value_unset(pointerof(out_string))
```
Comment (hugopl) [inline spec/inheritance_spec.cr:155]: You can also avoid the GValue mess here:
```
      value = uninitialized Pointer(LibC::Char)
      LibGObject.g_object_get(obj, "str-ing", pointerof(value), Pointer(Void).null)
      String.new(value).should eq("...")
```

File: src/bindings/g_object/value.cr [hunk 4]
```diff
@@ -38,6 +41,12 @@ module GObject
       when UInt8         then LibGObject.g_value_set_uchar(ptr, value)
       when GLib::Variant then LibGObject.g_value_set_variant(ptr, value)
       when ParamSpec     then LibGObject.g_value_set_param(ptr, value)
+      when Enum
+        if value.class._is_flags_enum?
+          LibGObject.g_value_set_flags(ptr, value)
+        else
+          LibGObject.g_value_set_enum(ptr, value)
+        end
```
Comment (hugopl) [inline src/bindings/g_object/value.cr:49]: Correct me if I'm wrong, but I believe you added this change due to the use of GValue to read/write properties, it's a good change, but would be nice to go in a separate commit with their own set of tests.

File: src/bindings/g_object/enum.cr [hunk 5]
```diff
@@ -35,8 +35,13 @@ abstract struct Enum
     @@_g_type
   end
 
+  # :nodoc:
+  def self._is_flags_enum? : Bool
+    {{ !!@type.annotation(Flags) }}
+  end
```
Comment (hugopl) [inline src/bindings/g_object/enum.cr:41]: This could be `GICrystal::flag?(obj)` instead.

File: src/bindings/g_object/object.cr [hunk 6]
```diff
@@ -33,6 +39,290 @@ module GObject
...
+        def unsafe_do_set_property(property_id : UInt32, gvalue : Void*, param_spec : Void*) : Nil
+          {% verbatim do %}
+            {% begin %}
+              {% instance_vars = @type.instance_vars.select(&.annotation(GObject::Property)) %}
+
+              case property_id
+              {% for var, i in instance_vars %}
+                {% if @type.has_method?("#{var.name}=") %}
+                  when {{ i + 1 }}
+                    raw = GObject::Value.raw(GObject.fundamental_g_type({{ var.type }}), gvalue)
```
Comment (hugopl) [inline src/bindings/g_object/object.cr:97]: GValue doesn't need to be used here neither, so one less heap allocation :tada:

**Review by hugopl — CHANGES_REQUESTED** [review]

**Comment by hugopl** [comment]
> Should we just always assume it's readable?

If it's possible to raise a compile error if there's no setter or getter it would be even greater.

File: src/bindings/g_object/object.cr (hunk 6 above)
Comment (BlobCodes) [reply src/bindings/g_object/object.cr:97]: I think there is no heap allocation here, I added a class method.

File: spec/inheritance_spec.cr (hunk 2 above)
Comment (BlobCodes) [reply spec/inheritance_spec.cr:111]: In "can set GObject properties" we check if we correctly implemented `unsafe_do_set_property` and in `can get GObject properties` we check if we correctly implemented `unsafe_do_get_property`. I think swapping it would be confusing.

Comment (BlobCodes) [reply spec/inheritance_spec.cr:111]: > Better to create a properties_spec.cr file with the properties spec instead of put the tests in inheritance specs.
There is already a `properties_spec.cr` and I don't know if this really fits there.

File: src/bindings/g_object/value.cr (hunk 4 above)
Comment (BlobCodes) [reply src/bindings/g_object/value.cr:49]: I think this change is small enough to be inside this PR.

File: src/bindings/g_object/enum.cr (hunk 5 above)
Comment (BlobCodes) [reply src/bindings/g_object/enum.cr:41]: No, in the GValue implementation we need to find out if a given class object is a flags enum (`value.class._is_flags_enum?`), but we can't access annotations at runtime

**Comment by hugopl** [comment]
This MR is on my queue, I believe this week I’ll have time for it.

File: src/bindings/g_object/type.cr [hunk 7]
```diff
@@ -22,6 +22,144 @@ module GObject
...
+      LibGObject.g_param_spec_flags(name, nick, blurb, klass.g_type, default || 0, flags)
+    else
+      LibGObject.g_param_spec_enum(name, nick, blurb, klass.g_type, default || 0, flags)
+    end
+  end
+
+  # :nodoc:
+  def self.fundamental_g_type(klass : String.class) : UInt64
+    TYPE_STRING
+  end
```
Comment (hugopl) [inline src/bindings/g_object/type.cr:93]: In `type.cr` we define a `self.g_type` method for most of the types, I think we can use them.

Anyway, I think this is good to be merged anyway and minor improvements can land in future patches.

Awesome work!!

**Review by hugopl — APPROVED** [review]

**Comment by hugopl** [comment]
I rebased the code locally here to solve the conflicts (just a minor one on Property annotation declaration), but then the tests failed

```
..............................................................*.......**................................................**.............*..............................
(process:4250): GLib-GObject-CRITICAL **: 09:20:10.057: validate_and_install_class_property: assertion 'class->set_property != NULL' failed

(process:4250): GLib-GObject-CRITICAL **: 09:20:10.058: validate_and_install_class_property: assertion 'class->set_property != NULL' failed

(process:4250): GLib-GObject-CRITICAL **: 09:20:10.058: validate_and_install_class_property: assertion 'class->set_property != NULL' failed

(process:4250): GLib-GObject-CRITICAL **: 09:20:10.058: validate_and_install_class_property: assertion 'class->set_property != NULL' failed

(process:4250): GLib-GObject-CRITICAL **: 09:20:10.058: validate_and_install_class_property: assertion 'class->set_property != NULL' failed

(process:4250): GLib-GObject-WARNING **: 09:20:10.058: g_object_set_is_valid_property: object class 'UserObjectWithGProperties' has no property named 'str-ing'
F
(process:4250): GLib-GObject-WARNING **: 09:20:10.059: g_object_get_is_valid_property: object class 'UserObjectWithGProperties' has no property named 'str-ing'
F................

Pending:
  GVariant can be used in return values
  GObject vfuncs can have return values
  GObject vfuncs can be from objects, not interfaces
  Raw C arrays of filenames can be received as argument as Array(Path)
  Raw C arrays of filenames can be received as argument as Tuple(Path)
  GObject signals test emit signals with return values

Failures:

  1) Classes inheriting GObject::Object can set GObject properties
     Failure/Error: obj.str_ing.should eq("test value")

       Expected: "test value"
            got: "default"

     # spec/inheritance_spec.cr:115

  2) Classes inheriting GObject::Object can get GObject properties
     Failure/Error: String.new(out_string).should eq("default")

       Expected: "default"
            got: "\xA7\u0001"

     # spec/inheritance_spec.cr:133

Finished in 106.54 milliseconds
184 examples, 2 failures, 0 errors, 6 pending
```

**Comment by BlobCodes** [comment]
I just fetched master and it works.

**Comment by hugopl** [comment]
Found the cause... it's the lack of GObject::virtual annotation on get_property virtual method... merging...

**Comment by BlobCodes** [comment]
> Found the cause... it's the lack of GObject::virtual annotation on get_property virtual method... merging...

No, you're using an old version then

File: src/bindings/g_object/object.cr [hunk 8]
```diff
@@ -67,6 +73,296 @@ module GObject
...
+                pspec = GObject.create_param_spec({{ var.type }}, name, nick, blurb, flags, {{ other_args.map { |tuple| "#{tuple[0]}: #{tuple[1]}".id }.splat }})
+                @@_g_param_specs[{{ i }}] = pspec.as(LibGObject::ParamSpec*)
+                LibGObject.g_object_class_install_property(klass, {{ i + 1 }}, pspec)
+              {% end %}
+            {% end %}
+          {% end %}
+        end
+
+        # :nodoc:
+        @[GObject::Virtual(name: "get_property", unsafe: true)]
```
Comment (BlobCodes) [inline src/bindings/g_object/object.cr:108]: Here

Comment (hugopl) [reply src/bindings/g_object/object.cr:108]: Weird... I got your last version with the merge from master from minutes ago... but after the rebase the git auto-merge must have messed up thing in object.cr file.

And now I ask what was really merged :rofl: , anyway... I reviewed the diff and it looks the same.
